pkg/agent/client/nodeconn.go

```diff
@@ -0,0 +1,44 @@
+package client
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"google.golang.org/grpc"
+)
+
+type nodeConn struct {
+	conn               *grpc.ClientConn
+	refcount           int32
+	mu                 sync.RWMutex
+	isMarkedForRelease bool
+}
+
+func newNodeConn(conn *grpc.ClientConn) *nodeConn {
+	return &nodeConn{
+		conn:     conn,
+		refcount: 1,
+	}
+}
+
+func (c *nodeConn) AddRef() {
+	atomic.AddInt32(&c.refcount, 1)
+}
+
+func (c *nodeConn) Release() {
+	if atomic.AddInt32(&c.refcount, -1) == 0 {
+		c.conn.Close()
+	}
+}
+
+func (c *nodeConn) MarkForRelease() {
```

Comment:
MarkForRelease is really a way to say "the client has no more interest in this connection", as such it should be in charge or releasing the implicit ref taken by `newNodeConn` (but only once). I'd suggest maybe something like:

```
func (c *nodeConn) MarkDead() {
    c.mu.Lock()
    if c.alive {
        c.release()
        c.alive = false
    }
    c.mu.Unlock()
}
```

`release` does the decrement and close if needed. Release() takes the lock and calls release().

Now that we need atomicity over multiple variables, I'd drop the atomic ops and just use a mutex.

The benefit of this approach is that `client.Release()` can just call `nodeConn.Release()`